valkyrietry.go

```diff
@@ -0,0 +1,253 @@
+package valkyrietry
+
+import (
+	"context"
+	"math/rand"
+	"time"
+)
+
+// RetryFunc
+// Implement this function for retryable actions within the `Do` or `DoWithContext` functions.
+// It is necessary to define this function if you want to utilize the
+// retry mechanism inside `Do` or `DoWithContext`.
+type RetryFunc func() error
+
+// RetryFuncWithData
+// Implement this function for retryable actions within the `DoWithData` or `DoWithDataAndContext` functions.
+// It is necessary to define this function if you intend to use
+// the retry mechanism inside `DoWithData` or `DoWithDataAndContext`.
+type RetryFuncWithData[T any] func() (T, error)
+
+type Valkyrietry struct {
+	Configuration *Configuration
+
+	ctx context.Context
+}
+
+func defaultValkyrietry(options ...Option) *Valkyrietry {
+	defaultConfiguration := &Configuration{
+		MaxRetryAttempts:       DefaultMaxRetryAttempt,
+		InitialRetryDelay:      DefaultRetryDelay,
+		RetryBackoffMultiplier: DefaultRetryBackoffMultiplier,
+		JitterPercentage:       DefaultJitter,
+	}
+
+	defaultValue := &Valkyrietry{
+		Configuration: defaultConfiguration,
+		ctx:           context.Background(),
+	}
+
+	for _, opt := range options {
+		opt(defaultConfiguration)
+	}
+
+	return defaultValue
+}
+
+func defaultValkyrietryWithContext(ctx context.Context, options ...Option) *Valkyrietry {
+	defaultConfiguration := &Configuration{
```

Comment:
**suggestion (non-blocking)**: wondering whether we could just clone default configuration in here

Reply:
Ah, you're right.t! I'll make that update.